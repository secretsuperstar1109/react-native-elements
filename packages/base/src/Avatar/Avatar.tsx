import React from 'react';
import {
  View,
  Text,
  Image as RNImage,
  StyleSheet,
  Pressable,
  StyleProp,
  ViewStyle,
  TextStyle,
  ImageSourcePropType,
  ImageStyle,
} from 'react-native';
import {
  InlinePressableProps,
  renderNode,
  RneFunctionComponent,
} from '../helpers';
import { IconObject, Icon } from '../Icon';
import { ImageProps, Image } from '../Image';

export const avatarSizes = {
  small: 34,
  medium: 50,
  large: 75,
  xlarge: 150,
};

type AvatarIcon = IconObject & {
  iconStyle?: StyleProp<TextStyle>;
};

export interface AvatarProps extends InlinePressableProps {
  /** Component for enclosing element (eg: TouchableHighlight, View, etc).
   *
   *  @default `Press handlers present then Pressable else View`
   */
  Component?: typeof React.Component;

  /** Callback function when pressing component. */
  onPress?(): void;

  /** Callback function when long pressing component. */
  onLongPress?(): void;

  /** Styling for outer container.
   *
   * @default undefined
   */
  containerStyle?: StyleProp<ViewStyle>;

  /** Image source to be displayed on avatar.
   *
   * @default undefined
   */
  source?: ImageSourcePropType;

  /** Style for avatar image.
   *
   * @default undefined
   */
  avatarStyle?: ImageStyle;

  /** Makes the avatar circular.
   *
   * @default false
   */
  rounded?: boolean;

  /** Renders title in the placeholder.
   *
   * @default undefined
   */
  title?: string;

  /** Style for the title.
   *
   * @default undefined
   */
  titleStyle?: StyleProp<TextStyle>;

  /** Style for the view outside image or icon.
   *
   * @default undefined
   */
  overlayContainerStyle?: StyleProp<TextStyle>;

  /** Displays an icon as the main content of the Avatar. **Cannot be used alongside title**. When used with the `source` prop it will be used as the placeholder.
   *
   * @default undefined
   */
  icon?: AvatarIcon;

  /** Extra styling for icon component.
   *
   * @default undefined
   */
  iconStyle?: StyleProp<TextStyle>;

  /** Size of the avatar.
   *
   * @default small
   */
  size?: ('small' | 'medium' | 'large' | 'xlarge') | number;

  /** Custom element inside the avatar (by default, it's the title).
   *
   * @default undefined
   */
  renderCustomContent?: React.ReactElement<{}>;

  /** Optional properties to pass to the avatar e.g "resizeMode".
   *
   * @default undefined
   */
  imageProps?: Partial<ImageProps>;

  /** Custom ImageComponent for Avatar.
   *
   * @default undefined
   */
  ImageComponent?: React.ComponentClass;
}

const AvatarTitle = ({
  size,
  title,
  titleStyle,
}: Pick<AvatarProps, 'title' | 'titleStyle' | 'size'>) => {
  const width =
    typeof size === 'number' ? size : avatarSizes[size] || avatarSizes.small;
  const titleSize = width / 2;

  return (
    <Text
      style={StyleSheet.flatten([
        styles.title,
        { fontSize: titleSize },
        titleStyle,
      ])}
    >
      {title}
    </Text>
  );
};

const AvatarIcon = ({
  icon,
  iconStyle,
  size,
}: Pick<AvatarProps, 'icon' | 'iconStyle' | 'size'>) => {
  const width =
    typeof size === 'number' ? size : avatarSizes[size] || avatarSizes.small;
  const iconSize = width / 2;

  return (
    <Icon
      style={StyleSheet.flatten([iconStyle])}
      color={icon.color || 'white'}
      name={icon.name || 'account'}
      size={icon.size || iconSize}
      type={icon.type || 'material-community'}
    />
  );
};

/**
 * Avatars are found all over ui design from lists to profile screens.
 * They are commonly used to represent a user and can contain photos, icons, or even text.
 * */
export const Avatar: RneFunctionComponent<AvatarProps> = ({
  onPress,
  onLongPress,
  onPressIn,
  onPressOut,
  Component = onPress || onLongPress || onPressIn || onPressOut
    ? Pressable
    : View,
  containerStyle,
  icon,
  iconStyle,
  source,
  size = 'small',
  avatarStyle,
  rounded,
  title,
  titleStyle,
  overlayContainerStyle,
  imageProps,
  renderCustomContent,
  ImageComponent = RNImage,
  children,
  pressableProps,
  ...rest
}) => {
  const width =
    typeof size === 'number' ? size : avatarSizes[size] || avatarSizes.small;
  const height = width;

  const imageContainerStyle = StyleSheet.flatten([
    styles.overlayContainer,
    rounded && { borderRadius: width / 2, overflow: 'hidden' },
    overlayContainerStyle,
    imageProps && imageProps.containerStyle,
  ]);

  if (imageProps && imageProps.containerStyle) {
    delete imageProps.containerStyle;
  }

  let componentToRender;

  if (source) {
    componentToRender = (
      <Image
        testID="RNE__Avatar__Image"
        containerStyle={imageContainerStyle as StyleProp<TextStyle>}
        source={source}
        PlaceholderContent={
          (title && (
            <AvatarTitle title={title} titleStyle={titleStyle} size={size} />
          )) ||
          (icon && <AvatarIcon icon={icon} iconStyle={iconStyle} size={size} />)
        }
        placeholderStyle={StyleSheet.flatten([
          styles.placeholderStyle,
          imageProps && imageProps.placeholderStyle,
        ])}
        borderRadius={rounded ? width / 2 : undefined}
        {...imageProps}
        style={StyleSheet.flatten([
          styles.avatar,
          imageProps && imageProps.style,
          avatarStyle,
        ])}
        ImageComponent={ImageComponent}
      />
    );
  } else if (title) {
    componentToRender = (
      <AvatarTitle title={title} titleStyle={titleStyle} size={size} />
    );
  } else if (icon) {
    componentToRender = (
      <AvatarIcon icon={icon} iconStyle={iconStyle} size={size} />
    );
  } else if (renderCustomContent) {
    componentToRender = renderNode(undefined, renderCustomContent);
  }

  return (
    <Component
      style={StyleSheet.flatten([
        styles.container,
        { height, width },
        rounded && { borderRadius: width / 2 },
        containerStyle,
      ])}
      {...{
        onPress,
        onLongPress,
        onPressIn,
        onPressOut,
        ...pressableProps,
        ...rest,
      }}
    >
      {componentToRender}
      {children}
    </Component>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
  },
  avatar: {
    flex: 1,
  },
  overlayContainer: {
    flex: 1,
  },
  title: {
    color: '#ffffff',
    backgroundColor: 'transparent',
    textAlign: 'center',
    zIndex: 1,
  },
  placeholderStyle: {
    backgroundColor: 'transparent',
  },
});

Avatar.displayName = 'Avatar';
