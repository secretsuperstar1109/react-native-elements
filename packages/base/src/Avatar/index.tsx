import { Avatar as AvatarBase, AvatarProps } from './Avatar';
import { Accessory, AccessoryProps } from './Avatar.Accessory';

const Avatar = Object.assign(AvatarBase, {
  Accessory,
});
export { Avatar };

export type { AccessoryProps, AvatarProps };
