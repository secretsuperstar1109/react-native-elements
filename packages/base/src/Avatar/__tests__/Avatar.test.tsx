import React from 'react';
import { Avatar } from '..';
import { Pressable, View, Image, Text } from 'react-native';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { avatarSizes } from './../Avatar';
import { Icon } from '../../Icon';

describe('Avatar Component', () => {
  jest.useFakeTimers();

  it('should match snapshot', () => {
    const { toJSON } = renderWithWrapper(
      <Avatar source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should renders rounded', () => {
    const { wrapper } = renderWithWrapper(
      <Avatar source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} rounded />
    );
    expect(wrapper.findByType(View).props.style.borderRadius).toBe(
      avatarSizes.small / 2
    );
  });

  it('should allows custom imageProps', () => {
    const { wrapper } = renderWithWrapper(
      <Avatar
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        imageProps={{
          resizeMode: 'contain',
          containerStyle: { backgroundColor: 'red' },
        }}
      />
    );
    expect(wrapper.findByType(Image).props.resizeMode).toBe('contain');
  });

  it('should renders touchable if onPress given', () => {
    const { wrapper } = renderWithWrapper(
      <Avatar
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        onPress={() => null}
      />
    );
    expect(wrapper.findByType(Pressable)).toBeTruthy();
  });

  it('should render using image with imageProps', () => {
    const { wrapper } = renderWithWrapper(
      <Avatar
        source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
        imageProps={{ style: { backgroundColor: 'red' } }}
      />
    );
    const img = wrapper.findByType(Image);
    expect(img.props.source).toMatchObject({
      uri: 'https://i.imgur.com/0y8Ftya.jpg',
    });
    expect(img.props.style).toMatchObject({
      backgroundColor: 'red',
    });
  });

  describe('Sizes', () => {
    it.each`
      size
      ${'small'}
      ${'medium'}
      ${'large'}
      ${'xlarge'}
    `('accepts $size', ({ size }) => {
      const { wrapper } = renderWithWrapper(
        <Avatar
          source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
          size={size}
        />
      );
      expect(wrapper.findByType(View).props.style).toMatchObject({
        height: avatarSizes[size],
        width: avatarSizes[size],
      });
    });

    it('should accepts a number as size', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} size={30} />
      );
      expect(wrapper.findByType(View).props.style).toMatchObject({
        height: 30,
        width: 30,
      });
    });

    it('should have default size', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }} />
      );
      expect(wrapper.findByType(View).props.style).toMatchObject({
        height: avatarSizes.small,
        width: avatarSizes.small,
      });
    });
  });

  describe('Title Avatars', () => {
    it('should renders title if given', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar
          source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
          title="John"
        />
      );
      expect(wrapper.findByType(Text).props.children).toBe('John');
    });
  });

  describe('Icon Avatars', () => {
    it('should renders using icon prop', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar
          source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
          icon={{
            name: 'home',
            type: 'material-community',
          }}
          iconStyle={{
            backgroundColor: 'red',
          }}
        />
      );
      expect(wrapper.findByType(Icon).props).toMatchObject({
        name: 'home',
        type: 'material-community',
        style: { backgroundColor: 'red' },
      });
    });

    it('should render using icon with defaults', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar
          source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
          iconStyle={{
            backgroundColor: 'red',
          }}
          icon={{}}
        />
      );
      expect(wrapper.findByType(Icon).props).toMatchObject({
        name: 'account',
        color: 'white',
        style: { backgroundColor: 'red' },
      });
    });
  });

  describe('renderCustomContent Avatar', () => {
    it('should renders custom renderCustomContent', () => {
      const { wrapper } = renderWithWrapper(
        <Avatar
          source={{ uri: 'https://i.imgur.com/0y8Ftya.jpg' }}
          renderCustomContent={<Text testID="my-custom-placeholder">Hey</Text>}
        />,
        'my-custom-placeholder'
      );
      expect(wrapper.props.children).toBe('Hey');
    });
  });
});
