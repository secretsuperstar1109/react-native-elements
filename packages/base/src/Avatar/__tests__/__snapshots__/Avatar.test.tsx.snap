// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "height": 34,
        "justifyContent": "center",
        "width": 34,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        {
          "backgroundColor": "transparent",
          "flex": 1,
          "overflow": "hidden",
          "position": "relative",
        }
      }
    >
      <Image
        onLoad={[Function]}
        source={
          {
            "uri": "https://i.imgur.com/0y8Ftya.jpg",
          }
        }
        style={
          {
            "bottom": 0,
            "flex": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="RNE__Avatar__Image"
        transitionDuration={360}
      />
      <View
        accessibilityElementsHidden={true}
        collapsable={false}
        importantForAccessibility="no-hide-descendants"
        pointerEvents="none"
        style={
          {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#bdbdbd",
              "flex": 1,
              "justifyContent": "center",
            }
          }
          testID="RNE__Image__placeholder"
        />
      </View>
      <View
        style={
          {
            "flex": 1,
          }
        }
        testID="RNE__Image__children__container"
      />
    </View>
  </View>
</View>
`;
